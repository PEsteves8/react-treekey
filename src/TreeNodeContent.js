import React from 'react';

import { defaultHeader } from './TreeViewDefaultTemplates';

export default class TreeListItemContent extends React.Component {

    renderNodeContent() {
        let { template, node } = this.props;
        return template ? template(node) : defaultHeader(node);
    }

    render() {
        return this.renderNodeContent();
    }
}