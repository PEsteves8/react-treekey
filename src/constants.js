export const KEYS = { 
    LEFT: 'ArrowLeft',
    RIGHT: 'ArrowRight',
    UP: 'ArrowUp',
    DOWN: 'ArrowDown',
};
