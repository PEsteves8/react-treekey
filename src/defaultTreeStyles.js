export default {
  root: {
    outline: "none",
    paddingLeft: "20px",
    listStyle: "none",
  },
  node: {
    default: {
      paddingLeft: "100%",
      marginLeft: "-100%",
    },
    selected: {
      backgroundColor: "grey",
    },
  },
  nodeToggleWrapper: {
    display: "inline-block",
    padding: "0 5px",
    marginLeft: "-18px",
  },
  nodeToggleSvg: {
    points: "0,0,0,10,8,5",
    height: 10,
    width: 8,
  },
  nestedList: {
    listStyle: "none",
    paddingLeft: "8px",
  },
};
