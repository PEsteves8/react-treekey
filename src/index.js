export { TreeKey } from "./TreeKey.js";
