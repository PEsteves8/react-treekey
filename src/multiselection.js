export function multiselection() {
  let anchorNode;
  let lastSelectedNode;

  function findCommonParent(firstNode, secondNode) {
    if (firstNode === secondNode) {
      return [firstNode.$parent];
    }

    let nodeHierarchy = [firstNode];
    let otherNodeHierarchy = [secondNode];
    let currentNode;

    currentNode = firstNode;
    while (currentNode.$parent) {
      nodeHierarchy.push(currentNode.$parent);
      currentNode = currentNode.$parent;
    }

    currentNode = secondNode;
    while (currentNode.$parent) {
      otherNodeHierarchy.push(currentNode.$parent);
      currentNode = currentNode.$parent;
    }

    nodeHierarchy = nodeHierarchy.reverse();
    otherNodeHierarchy = otherNodeHierarchy.reverse();

    let maxLength = Math.max(nodeHierarchy.length, otherNodeHierarchy.length);

    // Once a different parent is found, return the previous one that was the same
    for (let i = 0; i < maxLength; i++) {
      if (nodeHierarchy[i] !== otherNodeHierarchy[i]) {
        return nodeHierarchy[i - 1];
      }
    }
  }

  function getRange(commonParent, firstNode, secondNode, expandedNodes) {
    if (firstNode === secondNode) {
      return [firstNode];
    }

    let currentNode = commonParent;

    let firstIndex = null;
    let secondIndex = null;
    let result = [];
    while (currentNode && (firstIndex === null || secondIndex === null)) {
      if (currentNode === firstNode) {
        firstIndex = 1;
      }

      if (currentNode === secondNode) {
        secondIndex = 1;
      }

      if (firstIndex || secondIndex) {
        result.push(currentNode);
      }

      if (currentNode.$children && expandedNodes.has(currentNode)) {
        currentNode = currentNode.$firstChild;
      } else if (currentNode.$nextNode) {
        currentNode = currentNode.$nextNode;
      } else {
        currentNode = currentNode.$parent.$nextNode;
      }
    }

    return result;
  }

  function removeNodeFromSelectedNodes(selectedNodes, nodesToRemove) {
    return selectedNodes.reduce((result, node) => {
      if (!nodesToRemove.includes(node)) {
        result.push(node);
      }
      return result;
    }, []);
  }

  function addNodeToSelectedNodes(selectedNodes, nodesToAdd) {
    return nodesToAdd.reduce(
      (result, node) => {
        if (!selectedNodes.includes(node)) {
          result.push(node);
        }
        return result;
      },
      [...selectedNodes]
    );
  }

  function getNodesUsingKeyModifiers(
    node,
    e,
    selectedNodes,
    expandedNodes
  ) {
    if (e && e.ctrlKey) {
      if (e.type === "click") {
        const result = !selectedNodes.includes(node)
          ? addNodeToSelectedNodes(selectedNodes, [node])
          : removeNodeFromSelectedNodes(selectedNodes, [node]);

        // anchor node is used as a reference for shift range selections
        anchorNode = result.length > 0 ? node : undefined;

        // last selected node is used as the previously clicked node within range selections
        lastSelectedNode = undefined;

        return result;
      }
    } else if (e && e.shiftKey) {
      let result = [...selectedNodes];

      // First deselect the nodes that should be deselected if any
      if (lastSelectedNode) {
        let commonParent = findCommonParent(node, lastSelectedNode);

        let nodeRange = getRange(
          commonParent,
          node,
          lastSelectedNode,
          expandedNodes
        );

        result = removeNodeFromSelectedNodes(result, nodeRange);
      }

      // Then select the whole range from the anchor node to the clicked node
      anchorNode = anchorNode || selectedNodes[0];
      {
        let commonParent = findCommonParent(node, anchorNode);
        let nodeRange = getRange(commonParent, node, anchorNode, expandedNodes);
        result = addNodeToSelectedNodes(result, nodeRange);
      }

      lastSelectedNode = node;
      return result;
    } else {
      let result = [node];
      anchorNode = node;
      lastSelectedNode = undefined;
      return result;
    }
  };

  return {
    getNodesUsingKeyModifiers,
  }
}
