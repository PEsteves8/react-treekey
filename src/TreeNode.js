import React from "react";

import TreeNodeContent from "./TreeNodeContent";
import TreeNodeToggle from "./TreeNodeToggle";

class TreeNode extends React.Component {
  constructor(props) {
    super(props);
    this.onClick = this.onClick.bind(this);
  }

  onClick(e) {
    e.stopPropagation();
    this.props.selectNewNode(this.props.node, e);
  }

  render() {
    const { style } = this.props;
    const isSelectedNode = this.props.selectedNodes.includes(this.props.node);

    let nodeTextStyles = {};

    if (style) {
      nodeTextStyles = { ...style.node.default };

      if (isSelectedNode) {
        nodeTextStyles = { ...nodeTextStyles, ...style.node.selected };
      }
    }
    const isExpanded = this.props.expandedNodes.includes(this.props.node);
    const templates = this.props.templates || {};

    return (
      <li
        onClick={this.onClick}
        className={`${isSelectedNode ? "treeview-selected-node" : ""}`}
      >
        <div style={nodeTextStyles}>
          {this.props.node.children && (
            <TreeNodeToggle
              template={templates.toggle}
              setToggling={this.props.setToggling}
              node={this.props.node}
              isExpanded={isExpanded}
            />
          )}
          <TreeNodeContent
            template={templates.header}
            node={this.props.node}
            iconsEnabled={this.props.iconsEnabled}
          />
        </div>

        {this.props.node.children && isExpanded && (
          <ul style={this.props.style.nestedList}>
            {this.props.node.children.map((item, idx) => {
              return (
                <TreeNode
                  style={this.props.style}
                  node={item}
                  key={idx}
                  selectNewNode={this.props.selectNewNode}
                  setToggling={this.props.setToggling}
                  iconsEnabled={this.props.iconsEnabled}
                  templates={templates}
                  selectedNode={this.props.selectedNode}
                  selectedNodes={this.props.selectedNodes}
                  expandedNodes={this.props.expandedNodes}
                />
              );
            })}
          </ul>
        )}
      </li>
    );
  }
}

export default TreeNode;
