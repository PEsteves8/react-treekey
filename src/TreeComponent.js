import React from "react";
import PropTypes from "prop-types";

import TreeNode from "./TreeNode";
import defaultTreeStyles from "./defaultTreeStyles";

import { setTreeInternalProperties } from "./TreeViewHelpers";

import { multiselection } from "./multiselection";

export class TreeKey extends React.Component {
  constructor(props) {
    super(props);

    // adds non enumerable properties to each node
    // to keep track of its relationships for keyboard control
    // they'll be left out if the user stringifies the tree
    setTreeInternalProperties(props.tree);

    this.state = {
      tree: props.tree,
      selectedNode: null,
      selectedNodes: [],
      expandedNodes: this.props.expandedNodes || [],
    };

    this.handleOnKeyDown = this.handleOnKeyDown.bind(this);

    this.setToggling = this.setToggling.bind(this);
    this.selectNewNode = this.selectNewNode.bind(this);

    this.getNodesUsingKeyModifiers = multiselection().getNodesUsingKeyModifiers;
  }

  findFirstParentsNextNode(node) {
    if (!node.$parent) return;
    if (node.$parent && node.$parent.$nextNode) return node.$parent.$nextNode;
    else return this.findFirstParentsNextNode(node.$parent);
  }

  findLastCollapsedChild(node) {
    let lastCollapsedChild = node;
    while (
      lastCollapsedChild.$lastChild &&
      this.isNodeExpanded(lastCollapsedChild)
    ) {
      lastCollapsedChild = lastCollapsedChild.$lastChild;
    }
    return lastCollapsedChild;
  }

  isNodeExpanded(node) {
    return this.state.expandedNodes.includes(node);
  }

  selectPreviousNode(node, e) {
    if (
      node.$previousNode &&
      this.isNodeExpanded(node.$previousNode) &&
      node.$previousNode.$lastChild
    ) {
      const lastCollapsedChild = this.findLastCollapsedChild(
        node.$previousNode
      );
      this.selectNewNode(lastCollapsedChild, e);
    } else if (node.$previousNode) {
      this.selectNewNode(node.$previousNode, e);
    } else if (node.$parent) {
      this.selectNewNode(node.$parent, e);
    }
  }

  selectNextNode(node, e) {
    if (node.children && this.isNodeExpanded(node)) {
      this.selectNewNode(node.$firstChild, e);
    } else if (node.$nextNode) {
      this.selectNewNode(node.$nextNode, e);
    } else if (node.$parent) {
      let firstParentWithNextNode = this.findFirstParentsNextNode(node);
      if (firstParentWithNextNode)
        this.selectNewNode(firstParentWithNextNode, e);
    }
  }

  handleOnKeyDown(e) {
    let node =
      this.props.selectedNode ||
      this.state.selectedNode ||
      this.props.selectedNodes[0];
    let handlers = {
      ArrowUp: () => this.selectPreviousNode(node, e),
      ArrowDown: () => this.selectNextNode(node, e),
      ArrowLeft: () => this.setToggling(node, false),
      ArrowRight: () => this.setToggling(node, true),
    };

    if (handlers[e.key]) {
      e.preventDefault();
      handlers[e.key]();
    }
  }

  selectNewNode(node, e) {
    if (!this.props.selectedNode) {
      this.setState({ selectedNode: node });
    }

    if (this.props.multiSelection) {
      let selectedNodes = this.getNodesUsingKeyModifiers(
        node,
        e,
        this.props.selectedNodes,
        this.state.expandedNodes
      );
      this.props.onSelectNode(selectedNodes);
    } else {
      this.props.onSelectNode(node, e);
    }
  }

  setToggling(node, shouldBeExpanded = !this.isNodeExpanded(node)) {
    // if no value is provided, then simply invert the previous one
    // the value is provided when using the arrow keys
    let expanded = this.state.expandedNodes;
    let isExpanded = this.isNodeExpanded(node);
    let expandedNodes;
    if (shouldBeExpanded && !isExpanded) {
      expandedNodes = [...expanded, node];
    } else if (!shouldBeExpanded && isExpanded) {
      expandedNodes = expanded.filter((n) => n !== node);
    }
    if (expandedNodes) {
      this.setState({ expandedNodes });
    }
  }

  componentDidMount() {
    if (!this.isStartingSelectedNodeSet()) {
      // Set the root as the selected node
      this.selectNewNode(this.props.tree);
    }

    if (this.state.expandedNodes.length === 0) {
      this.setState({ expandedNodes: [this.props.tree] });
    }
  }

  isStartingSelectedNodeSet() {
    return (
      this.props.selectedNode ||
      (this.props.selectedNodes && this.props.selectedNodes.length > 0)
    );
  }

  render() {
    const style = this.props.styles || defaultTreeStyles;
    const templates = this.props.templates;

    const selectedNode = this.props.selectedNode || this.state.selectedNode;
    const selectedNodes = this.props.selectedNodes;

    return (
      <ul
        tabIndex={0}
        onKeyDown={this.handleOnKeyDown}
        className="treeview-root"
        style={style.root}
      >
        <TreeNode
          selectedNode={selectedNode}
          selectedNodes={selectedNodes}
          templates={templates}
          node={this.state.tree}
          selectNewNode={this.selectNewNode}
          setToggling={this.setToggling}
          style={style}
          expandedNodes={this.state.expandedNodes}
        />
      </ul>
    );
  }
}

TreeKey.propTypes = {
  tree: PropTypes.object,
  onSelectNode: PropTypes.func,
  selectedNode: PropTypes.object,
  selectedNodes: PropTypes.arrayOf(PropTypes.object),
  expandedNodes: PropTypes.arrayOf(PropTypes.object),
  templates: PropTypes.objectOf(PropTypes.func),
  multiSelection: PropTypes.bool,
};
