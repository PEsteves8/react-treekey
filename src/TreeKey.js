import React from "react";
import PropTypes from "prop-types";

import TreeNode from "./TreeNode";
import defaultTreeStyles from "./defaultTreeStyles";

import { setTreeInternalProperties } from "./TreeViewHelpers";

import { multiselection } from "./multiselection";

import { KEYS } from "./constants";

export class TreeKey extends React.Component {
  constructor(props) {
    super(props);
    let { selectedNodes, expandedNodes, tree } = props;

    // adds non enumerable properties to each node
    // to keep track of its relationships for keyboard control
    // they'll be left out if the user stringifies the tree
    setTreeInternalProperties(tree);

    // set root as selected / expanded is it hasn't been specified
    this.state = {
      tree,
      lastSelectedNode: (selectedNodes && selectedNodes[0]) || tree,
      selectedNodes: selectedNodes || [tree],
      expandedNodes: expandedNodes || [tree],
    };

    this.handleOnKeyDown = this.handleOnKeyDown.bind(this);

    this.updateExpandedNodes = this.updateExpandedNodes.bind(this);
    this.selectNewNode = this.selectNewNode.bind(this);
    this.handleToggle = this.handleToggle.bind(this);

    this.getNodesUsingKeyModifiers = multiselection().getNodesUsingKeyModifiers;
  }

  findFirstParentsNextNode(node) {
    if (!node.$parent) return;

    if (node.$parent && node.$parent.$nextNode) {
      return node.$parent.$nextNode;
    }

    return this.findFirstParentsNextNode(node.$parent);
  }

  findLastCollapsedChild(node) {
    let lastCollapsedChild = node;
    while (
      lastCollapsedChild.$lastChild &&
      this.isNodeExpanded(lastCollapsedChild)
    ) {
      lastCollapsedChild = lastCollapsedChild.$lastChild;
    }
    return lastCollapsedChild;
  }

  isNodeExpanded(node) {
    return this.state.expandedNodes.includes(node);
  }

  selectPreviousNode(node, e) {
    if (
      node.$previousNode &&
      this.isNodeExpanded(node.$previousNode) &&
      node.$previousNode.$lastChild
    ) {
      const lastCollapsedChild = this.findLastCollapsedChild(
        node.$previousNode
      );
      this.selectNewNode(lastCollapsedChild, e);
    } else if (node.$previousNode) {
      this.selectNewNode(node.$previousNode, e);
    } else if (node.$parent) {
      this.selectNewNode(node.$parent, e);
    }
  }

  selectNextNode(node, e) {
    if (node.children && this.isNodeExpanded(node)) {
      this.selectNewNode(node.$firstChild, e);
    } else if (node.$nextNode) {
      this.selectNewNode(node.$nextNode, e);
    } else if (node.$parent) {
      let firstParentWithNextNode = this.findFirstParentsNextNode(node);
      if (firstParentWithNextNode)
        this.selectNewNode(firstParentWithNextNode, e);
    }
  }

  handleOnKeyDown(e) {
    let node = this.state.lastSelectedNode;
    const { LEFT, RIGHT, UP, DOWN } = KEYS;
    let handlers = {
      [UP]: () => this.selectPreviousNode(node, e),
      [DOWN]: () => this.selectNextNode(node, e),
      [LEFT]: () => this.handleToggle(node, LEFT),
      [RIGHT]: () => this.handleToggle(node, RIGHT),
    };

    if (handlers[e.key]) {
      e.preventDefault();
      handlers[e.key]();
    }
  }

  selectNewNode(node, e) {
    let { multiSelection } = this.props;
    let selectedNodes = multiSelection
      ? this.getNodesUsingKeyModifiers(
          node,
          e,
          this.state.selectedNodes,
          this.state.expandedNodes
        )
      : [node];

    if (this.shouldAutoSelectNodes()) {
      this.setState({ selectedNodes });
    }

    this.setState({ lastSelectedNode: node });
    this.props.onSelectNode(multiSelection ? selectedNodes : selectedNodes[0]);
  }

  updateExpandedNodes(node, shouldExpand) {
    let { expandedNodes } = this.state;
    let result = shouldExpand
      ? [...expandedNodes, node]
      : expandedNodes.filter((n) => n !== node);
    return result;
  }

  handleToggle(node, key) {
    // the explicit value is only provided when using the arrow keys
    // with click, simply invert the previous value
    const { LEFT, RIGHT } = KEYS;
    let isExpanded = this.isNodeExpanded(node);

    let expandedNodes;

    if (key === LEFT) {
      if (
        !node.$previousNode &&
        node.$parent &&
        this.isNodeExpanded(node.$parent) &&
        !isExpanded
      ) {
        this.selectNewNode(node.$parent);
      } else if (isExpanded) {
        expandedNodes = this.updateExpandedNodes(node, false);
      }
    } else if (key === RIGHT && node.children) {
      if (!isExpanded) {
        expandedNodes = this.updateExpandedNodes(node, true);
      } else {
        this.selectNewNode(node.children[0]);
      }
    } else if (node.children) {
      // if click event
      expandedNodes = this.updateExpandedNodes(node, !isExpanded);
    }

    if (expandedNodes) {
      if (this.shouldAutoExpandNodes()) {
        this.setState({ expandedNodes });
      }

      if (this.props.onExpandNode) {
        this.props.onExpandNode(expandedNodes);
      }
    }
  }

  shouldAutoSelectNodes() {
    return !this.props.selectedNodes;
  }

  shouldAutoExpandNodes() {
    return !this.props.expandedNodes;
  }

  static getDerivedStateFromProps(props, state) {
    // if parents isn't sending these values explicitly
    // the component becomes controlled
    let selectedNodes = props.selectedNodes || state.selectedNodes;
    let expandedNodes = props.expandedNodes || state.expandedNodes;
    return { selectedNodes, expandedNodes };
  }

  render() {
    let style = this.props.styles || defaultTreeStyles;
    let templates = this.props.templates;

    let { selectedNodes, expandedNodes, tree } = this.state;

    return (
      <ul
        tabIndex={0}
        onKeyDown={this.handleOnKeyDown}
        className="treeview-root"
        style={style.root}
      >
        <TreeNode
          selectedNodes={selectedNodes}
          templates={templates}
          node={tree}
          selectNewNode={this.selectNewNode}
          handleToggle={this.handleToggle}
          style={style}
          expandedNodes={expandedNodes}
          indentValue={1}
        />
      </ul>
    );
  }
}

TreeKey.propTypes = {
  tree: PropTypes.object,
  onSelectNode: PropTypes.func,
  onExpandNode: PropTypes.func,
  selectedNodes: PropTypes.arrayOf(PropTypes.object),
  expandedNodes: PropTypes.arrayOf(PropTypes.object),
  templates: PropTypes.objectOf(PropTypes.func),
  multiSelection: PropTypes.bool,
};
