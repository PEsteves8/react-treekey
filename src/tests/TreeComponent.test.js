import React, { useState } from "react";
import { TreeKey } from "../TreeComponent";
import { treeA } from "../../examples/src/data";
import { render, unmountComponentAtNode } from "react-dom";
import { act } from "react-dom/test-utils";

const TreeKeyTestComponent = (props) => {
  const { selectedNodes, ...propsRest } = props;
  const [selectedNodesProp, setSelectNodesProp] = useState(selectedNodes);
  const onSelectNode = (nodes) => {
    setSelectNodesProp(nodes);
  };

  return (
    <TreeKey
      tree={treeA}
      onSelectNode={onSelectNode}
      selectedNodes={selectedNodesProp}
      {...propsRest}
    ></TreeKey>
  );
};

let container = null;
beforeEach(() => {
  // setup a DOM element as a render target
  container = document.createElement("div");
  document.body.appendChild(container);
});

afterEach(() => {
  // cleanup on exiting
  unmountComponentAtNode(container);
  container.remove();
  container = null;
});

describe("<TreeKey>", () => {
  test("should match snapshot", () => {
    const props = {
      selectedNodes: [treeA],
      expandedNodes: new Set([treeA]),
      multiSelection: true,
    };
    act(() => {
      render(<TreeKeyTestComponent {...props} />, container);
    });
    expect(container.innerHTML).toMatchSnapshot();
  });

  describe("handleKeyDown", () => {
    describe("when pressing up or down arrow keys", () => {
      test.each([
        ["ArrowUp", treeA.$children[0], treeA],
        ["ArrowUp", treeA.$children[1], treeA.$children[0]],
        ["ArrowUp", treeA.$children[0], treeA],
        // if previous sibling has children and is expanded
        [
          "ArrowUp",
          treeA.$children[3],
          treeA.$children[2].$children[4],
          new Set([treeA, treeA.$children[2]]),
        ],
        // if previous sibling has children but it's collapsed
        ["ArrowUp", treeA.$children[3], treeA.$children[2], new Set([treeA])],

        ["ArrowDown", treeA, treeA.$children[0], new Set([treeA])],
        ["ArrowDown", treeA.$children[0], treeA.$children[1], new Set([treeA])],
        [
          "ArrowDown",
          treeA.$children[2].$children[4],
          treeA.$children[3],
          new Set([treeA, treeA.$children[2]]),
        ],
        // if current node has children and is expanded
        [
          "ArrowDown",
          treeA.$children[2],
          treeA.$children[2].$children[0],
          new Set([treeA, treeA.$children[2]]),
        ],
        // if current node has children but is not expanded
        ["ArrowDown", treeA.$children[2], treeA.$children[3]],
      ])(
        "by pressing key %s, if starting node is %s next node is %s",
        (key, startNode, endNode, expandedNodes = new Set([treeA])) => {
          const props = {
            selectedNodes: [startNode],
            expandedNodes,
            multiSelection: true,
          };

          act(() => {
            render(<TreeKeyTestComponent {...props} />, container);
          });

          const treeEl = container.querySelector("ul");
          act(() => {
            treeEl.dispatchEvent(
              new KeyboardEvent("keydown", { bubbles: true, key })
            );
          });

          const selectedElements = document.querySelectorAll(
            ".treeview-selected-node > div > span"
          );
          expect(selectedElements.length).toBe(1);
          expect(selectedElements[0].innerHTML).toBe(endNode.name);
        }
      );

      test.each([
        ["ArrowUp", treeA],
        ["ArrowDown", treeA.$children[4]],
      ])(
        "if nodes are on the edges, by pressing key %s, if starting node is nothing changes",
        (key, startNode) => {
          const props = {
            selectedNodes: [startNode],
            expandedNodes: new Set([treeA]),
            multiSelection: true,
          };

          act(() => {
            render(<TreeKeyTestComponent {...props} />, container);
          });

          const selectedElements = document.querySelectorAll(
            ".treeview-selected-node > div > span"
          );

          expect(selectedElements.length).toBe(1);
          expect(selectedElements[0].innerHTML).toBe(startNode.name);

          const treeEl = container.querySelector("ul");
          act(() => {
            treeEl.dispatchEvent(
              new KeyboardEvent("keydown", { bubbles: true, key })
            );
          });

          expect(selectedElements.length).toBe(1);
          expect(selectedElements[0].innerHTML).toBe(startNode.name);
        }
      );

      test("when SHIFT modifier is true, the new node will be added to the previous instead of replacing it", () => {
        const props = {
          selectedNodes: [treeA.$children[0]],
          expandedNodes: new Set([treeA, treeA.$children[2]]),
          multiSelection: true,
        };

        act(() => {
          render(<TreeKeyTestComponent {...props} />, container);
        });
        let selectedElements = document.querySelectorAll(
          ".treeview-selected-node > div > span"
        );

        expect(selectedElements.length).toBe(1);
        expect(selectedElements[0].innerHTML).toBe("child1");

        function testKeyPress(key, expectedNames) {
          let treeEl = container.querySelector("ul");
          
          act(() => {
            treeEl.dispatchEvent(
              new KeyboardEvent("keydown", {
                bubbles: true,
                key,
                shiftKey: true,
              })
            );
          });

          let selectedElements = document.querySelectorAll(
            ".treeview-selected-node > div > span"
          );
          
          expect(selectedElements.length).toBe(expectedNames.length);
          for (let i = 0; i < selectedElements.length; i++) {
            expect(expectedNames).toContain(selectedElements[i].innerHTML);
          }
        }

        testKeyPress("ArrowDown", ["child1", "child2"]);
        testKeyPress("ArrowDown", ["child1", "child2", "child3"]);
        testKeyPress("ArrowDown", [
          "child1",
          "child2",
          "child3",
          "grandchild1",
        ]);
        testKeyPress("ArrowDown", [
          "child1",
          "child2",
          "child3",
          "grandchild1",
          "grandchild2"
        ]);
        testKeyPress("ArrowDown", [
          "child1",
          "child2",
          "child3",
          "grandchild1",
          "grandchild2",
          "grandchild3", // grandchild 2 is collapsed, so nextnode isn't a child, but a sibling
        ]);
        testKeyPress("ArrowUp", ["child1", "child2", "child3", "grandchild1", "grandchild2"]);
        testKeyPress("ArrowUp", ["child1", "child2", "child3", "grandchild1"]);

        selectedElements = document.querySelectorAll(".treeview-selected-node > div > span");
        act(() => {
          selectedElements[3].dispatchEvent(new MouseEvent("click", { bubbles: true }));
        });
        selectedElements = document.querySelectorAll(".treeview-selected-node > div > span");
        expect(selectedElements.length).toBe(1);
        expect(selectedElements[0].innerHTML).toBe("grandchild1");

        testKeyPress("ArrowUp", ["grandchild1", "child3"]);
        testKeyPress("ArrowUp", ["grandchild1", "child3", "child2"]);
      });
    });
  });

  describe("mouse click", () => {
    test("when there is neither CTRL nor SHIFT modifiers, select the clicked node", () => {
      const props = {
        selectedNodes: [treeA.$children[0]],
        expandedNodes: new Set([treeA, treeA.$children[2]]),
        multiSelection: true,
      };

      act(() => {
        render(<TreeKeyTestComponent {...props} />, container);
      });

      act(() => {
        document.querySelector('ul > li > ul > li:nth-child(3) > ul > li:nth-child(2)').dispatchEvent(new MouseEvent("click", { bubbles: true }));
      });

      let selectedElements = document.querySelectorAll(
        ".treeview-selected-node > div > span"
      );
      expect(selectedElements.length).toBe(1);
      expect(selectedElements[0].innerHTML).toBe(treeA.$children[2].$children[1].name);

      act(() => {
        document.querySelector('ul > li > ul > li:nth-child(2)').dispatchEvent(new MouseEvent("click", { bubbles: true }));
      });
      selectedElements = document.querySelectorAll(
        ".treeview-selected-node > div > span"
      );
      expect(selectedElements.length).toBe(1);
      expect(selectedElements[0].innerHTML).toBe(treeA.$children[1].name);
    });

    test("when there is CTRL modifier, is node is selected, remove it; otherwise add it", () => {
      const props = {
        selectedNodes: [treeA.$children[0]],
        expandedNodes: new Set([treeA, treeA.$children[2]]),
        multiSelection: true,
      };

      act(() => {
        render(<TreeKeyTestComponent {...props} />, container);
      });

      act(() => {
        document.querySelector('ul > li > ul > li:nth-child(3) > ul > li:nth-child(2)').dispatchEvent(new MouseEvent("click", { bubbles: true, ctrlKey: true }));
      });

      let selectedElements = document.querySelectorAll(
        ".treeview-selected-node > div > span"
      );
      expect(selectedElements.length).toBe(2);
      expect(selectedElements[0].innerHTML).toBe(treeA.$children[0].name);
      expect(selectedElements[1].innerHTML).toBe(treeA.$children[2].$children[1].name);

      act(() => {
        document.querySelector('ul > li > ul > li:nth-child(2)').dispatchEvent(new MouseEvent("click", { bubbles: true, ctrlKey: true }));
      });
      selectedElements = document.querySelectorAll(
        ".treeview-selected-node > div > span"
      );
      expect(selectedElements.length).toBe(3);
      expect(selectedElements[0].innerHTML).toBe(treeA.$children[0].name);
      expect(selectedElements[1].innerHTML).toBe(treeA.$children[1].name);
      expect(selectedElements[2].innerHTML).toBe(treeA.$children[2].$children[1].name);

      // clicking on currently selected node removes it from list
      act(() => {
        document.querySelector('ul > li > ul > li:nth-child(3) > ul > li:nth-child(2)').dispatchEvent(new MouseEvent("click", { bubbles: true, ctrlKey: true }));
      });
      selectedElements = document.querySelectorAll(".treeview-selected-node > div > span");
      expect(selectedElements.length).toBe(2);
      expect(selectedElements[0].innerHTML).toBe(treeA.$children[0].name);
      expect(selectedElements[1].innerHTML).toBe(treeA.$children[1].name);
    });

    test("when there is SHIFT modifier, using last node clicked before SHIFT pressed as anchor, add or remove nodes accordingly", () => {
      const props = {
        selectedNodes: [treeA.$children[2]],
        expandedNodes: new Set([treeA, treeA.$children[2]]),
        multiSelection: true,
      };

      act(() => {
        render(<TreeKeyTestComponent {...props} />, container);
      });

      act(() => {
        document.querySelector('.treeview-root > li > ul > li:nth-child(4)')
          .dispatchEvent(new MouseEvent("click", { bubbles: true, shiftKey: true }));
      });

      let selectedElements = document.querySelectorAll(
        ".treeview-selected-node > div > span"
      );
      let expectedNodes = [treeA.$children[2],
        treeA.$children[2].$children[0], treeA.$children[2].$children[1], treeA.$children[2].$children[2], treeA.$children[2].$children[3], treeA.$children[2].$children[4], treeA.$children[3]
      ];
      let expectedNames = expectedNodes.map(node => node.name);
      expect(selectedElements.length).toBe(expectedNames.length);
      for (let i = 0; i < selectedElements.length; i++) {
        expect(expectedNames).toContain(selectedElements[i].innerHTML);
      }

      
      act(() => {
        document.querySelector('ul > li > ul > li:nth-child(3) > ul > li:nth-child(2)')
          .dispatchEvent(new MouseEvent("click", { bubbles: true, shiftKey: true }));
      });
      selectedElements = document.querySelectorAll(
        ".treeview-selected-node > div > span"
      );
      expectedNodes = [treeA.$children[2],
        treeA.$children[2].$children[0], treeA.$children[2].$children[1]];
      expectedNames = expectedNodes.map(node => node.name);
      expect(selectedElements.length).toBe(expectedNames.length);
      for (let i = 0; i < selectedElements.length; i++) {
        expect(expectedNames).toContain(selectedElements[i].innerHTML);
      }



      act(() => {
        document.querySelector('ul > li')
          .dispatchEvent(new MouseEvent("click", { bubbles: true, shiftKey: true }));
      });
      selectedElements = document.querySelectorAll(
        ".treeview-selected-node > div > span"
      );
      expectedNodes = [treeA.$children[2],
        treeA.$children[1], treeA.$children[0], treeA];
      expectedNames = expectedNodes.map(node => node.name);
      expect(selectedElements.length).toBe(expectedNames.length);
      for (let i = 0; i < selectedElements.length; i++) {
        expect(expectedNames).toContain(selectedElements[i].innerHTML);
      }
    });
  });
});