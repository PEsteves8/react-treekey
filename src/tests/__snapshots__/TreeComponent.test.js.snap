// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TreeKey> should match snapshot 1`] = `"<ul tabindex=\\"0\\" class=\\"treeview-root\\" style=\\"outline: none; padding-left: 20px; list-style: none;\\"><li class=\\"treeview-selected-node\\"><div style=\\"padding-left: 100%; margin-left: -100%; background-color: grey;\\"><div style=\\"display: inline-block; padding: 0px 5px; margin-left: -18px;\\"><svg height=\\"10\\" width=\\"8\\"><polygon points=\\"0,0,0,10,8,5\\" style=\\"fill: white; stroke-width: 0; transform-origin: center; transform: rotate(90deg);\\"></polygon></svg></div><span>root</span></div><ul style=\\"list-style: none; padding-left: 8px;\\"><li class=\\"\\"><div style=\\"padding-left: 100%; margin-left: -100%;\\"><span>child1</span></div></li><li class=\\"\\"><div style=\\"padding-left: 100%; margin-left: -100%;\\"><span>child2</span></div></li><li class=\\"\\"><div style=\\"padding-left: 100%; margin-left: -100%;\\"><div style=\\"display: inline-block; padding: 0px 5px; margin-left: -18px;\\"><svg height=\\"10\\" width=\\"8\\"><polygon points=\\"0,0,0,10,8,5\\" style=\\"fill: white; stroke-width: 0; transform-origin: center; transform: rotate(0);\\"></polygon></svg></div><span>child3</span></div></li><li class=\\"\\"><div style=\\"padding-left: 100%; margin-left: -100%;\\"><span>child4</span></div></li><li class=\\"\\"><div style=\\"padding-left: 100%; margin-left: -100%;\\"><div style=\\"display: inline-block; padding: 0px 5px; margin-left: -18px;\\"><svg height=\\"10\\" width=\\"8\\"><polygon points=\\"0,0,0,10,8,5\\" style=\\"fill: white; stroke-width: 0; transform-origin: center; transform: rotate(0);\\"></polygon></svg></div><span>child5</span></div></li></ul></li></ul>"`;
